src/algebra/ordered_ring_hom.lean

```diff
@@ -0,0 +1,146 @@
+/-
+Copyright (c) 2021 Alex J. Best. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex J. Best
+-/
+import order.rel_iso
+import algebra.ordered_ring
+
+/-!
+# Ordered ring homomorphisms
+
+Homomorphisms between ordered rings that respect the ordering.
+
+## Main definitions
+
+* `ordered_ring_hom` : A homomorphism `f` between two `ordered_semiring`s that has the property that
+  `x ≤ y → f x ≤ f y`.
+
+## Main results
+
+* `ordered_ring_equiv_eq_cut_ordered_ring_equiv` : Uniqueness of the `ordered_ring_equiv`
+
+## Notation
+
+* `→+*o`: equivalence of ordered rings.
+
+## Tags
+ordered ring, equivalence, order homomorphism
+-/
+
+/-- Homomorphism commuting with multiplicative, additive and order structure. -/
+@[nolint has_inhabited_instance]
+structure ordered_ring_hom (R S : Type*) [ordered_semiring R] [ordered_semiring S]
+  extends R →+* S :=
+(map_rel' : ∀ {a b}, a ≤ b → to_fun a ≤ to_fun b)
+
+/-- Reinterpret an ordered ring homomorphism as a ring homomorphism. -/
+add_decl_doc ordered_ring_hom.to_ring_hom
+
+namespace ordered_ring_hom
+
+infix ` →+*o `:25 := ordered_ring_hom
+
+variables {R S : Type*} [ordered_semiring R] [ordered_semiring S]
+
+/-- Reinterpret an ordered ring homomorphism as an order homomorphism. -/
+def to_rel_hom (f : R →+*o S) : ((≤) : R → R → Prop) →r ((≤) : S → S → Prop) := { ..f }
+
+instance has_coe_to_ring_hom : has_coe (R →+*o S) (R →+* S) := ⟨to_ring_hom⟩
+instance has_coe_to_hom : has_coe (R →+*o S) (R → S) := ⟨ring_hom.to_fun ∘ to_ring_hom⟩
+instance has_coe_to_rel_hom : has_coe (R →+*o S) (((≤) : R → R → Prop) →r ((≤) : S → S → Prop)) :=
+⟨to_rel_hom⟩
+
+instance : has_coe_to_fun (R →+*o S) := ⟨_, λ f, f.to_fun⟩
+
+@[simp] lemma to_ring_hom_eq_coe {f : R →+*o S} : f.to_ring_hom = f := rfl
+@[simp] lemma to_order_iso_eq_coe {f : R →+*o S} : f.to_rel_hom = f := rfl
+@[simp] lemma coe_ring_hom_to_fun_eq_coe_fun {f : R →+*o S} : (f : R →+* S).to_fun = f := rfl
+@[simp] lemma coe_ring_hom_coe_fun_eq_coe_fun {f : R →+*o S} : ((f : R →+* S) : R → S) = f := rfl
+@[simp] lemma coe_rel_hom_to_fun_eq_to_equiv {f : R →+*o S} :
+  (f : ((≤) : R → R → Prop) →r ((≤) : S → S → Prop)).to_fun = f := rfl
+@[simp] lemma coe_rel_hom_coe_fun_eq_to_equiv {f : R →+*o S} :
+  ((f : ((≤) : R → R → Prop) →r ((≤) : S → S → Prop)) : R → S) = f := rfl
+@[simp]
+lemma coe_mul_equiv_to_fun_eq_coe_fun {f : R →+*o S} : ((f : R →+* S) : R →* S).to_fun = f := rfl
+@[simp]
+lemma coe_mul_equiv_coe_fun_eq_coe_fun {f : R →+*o S} : (((f : R →+* S) : R →* S) : R → S) = f :=
+rfl
+@[simp]
+lemma coe_add_equiv_to_fun_eq_coe_fun {f : R →+*o S} : ((f : R →+* S) : R →+ S).to_fun = f := rfl
+@[simp]
+lemma coe_add_equiv_coe_fun_eq_coe_fun {f : R →+*o S} : (((f : R →+* S) : R →+ S) : R → S) = f :=
+rfl
```

Comment:
I think you want to establish some simp-normal forms here, with
```suggestion
-- establish simp-normal forms
@[simp] lemma to_fun_eq_coe {f : R →+*o S} : f.to_fun = f := rfl
@[simp] lemma to_ring_hom_eq_coe {f : R →+*o S} : f.to_ring_hom = (f : R →+* S) := rfl
@[simp] lemma to_order_iso_eq_coe {f : R →+*o S} : f.to_rel_hom = (f : (((≤) : R → R → Prop) →r ((≤) : S → S → Prop))) := rfl

@[simp] lemma coe_ring_hom {f : R →+*o S} : (f : R →+* S) = f := rfl
@[simp] lemma coe_rel_hom {f : R →+*o S} :
  (f : ((≤) : R → R → Prop) →r ((≤) : S → S → Prop)) = f := rfl
```

My understanding is that `simp` can then do all the other reductions automatically